sdk/cosmosdb/cosmos/src/bulk/BulkStreamer.ts

```diff
@@ -0,0 +1,267 @@
+ * will be limited. Two callback implementations define how a particular request should be executed, and how operations should be retried. When the streamer dispatches a batch
+ * the batch will create a request and call the execute callback (executeRequest), if conditions are met, it might call the retry callback (reBatchOperation).
+ * @hidden
+ */
+
+export class BulkStreamer {
+  private readonly container: Container;
+  private readonly clientContext: ClientContext;
+  private readonly partitionKeyRangeCache: PartitionKeyRangeCache;
+  private readonly streamersByPartitionKeyRangeId: Map<string, BulkStreamerPerPartition>;
+  private readonly limitersByPartitionKeyRangeId: Map<string, Limiter>;
+  private options: RequestOptions;
+  private orderedResponse: BulkOperationResult[] = [];
+  private diagnosticNode: DiagnosticNodeInternal;
+  private operationPromises: Promise<BulkOperationResult>[] = [];
+  private operationIndex: number = 0;
+
+  /**
+   * @internal
+   */
+  constructor(
+    container: Container,
+    clientContext: ClientContext,
+    partitionKeyRangeCache: PartitionKeyRangeCache,
+  ) {
+    this.container = container;
+    this.clientContext = clientContext;
+    this.partitionKeyRangeCache = partitionKeyRangeCache;
+    this.streamersByPartitionKeyRangeId = new Map();
+    this.limitersByPartitionKeyRangeId = new Map();
+
+    this.executeRequest = this.executeRequest.bind(this);
+    this.reBatchOperation = this.reBatchOperation.bind(this);
+  }
+  /**
+   *
+   * @internal
+   */
+  initializeBulk(options: RequestOptions): void {
+    this.orderedResponse = [];
+    this.options = options;
+    this.operationIndex = 0;
+    this.operationPromises = [];
+    this.diagnosticNode = new DiagnosticNodeInternal(
+      this.clientContext.diagnosticLevel,
+      DiagnosticNodeType.CLIENT_REQUEST_NODE,
+      null,
+    );
+  }
+
+  /**
+   * adds operation(s) to the streamer
+   * @param operationInput - bulk operation or list of bulk operations
+   */
+  addOperations(operationInput: OperationInput | OperationInput[]): void {
+    if (Array.isArray(operationInput)) {
+      operationInput.forEach((operation) => {
+        const operationPromise = this.addOperation(operation);
+        this.operationPromises.push(operationPromise);
+      });
+    } else {
+      const operationPromise = this.addOperation(operationInput);
+      this.operationPromises.push(operationPromise);
+    }
+  }
+
+  private async addOperation(operation: OperationInput): Promise<BulkOperationResult> {
+    if (!operation) {
+      throw new ErrorResponse("Operation is required.");
+    }
+    const partitionKeyRangeId = await this.resolvePartitionKeyRangeId(operation);
+    const streamerForPartition = this.getOrCreateStreamerForPKRange(partitionKeyRangeId);
+    const retryPolicy = this.getRetryPolicy();
```

Comment:
why are we mapping retry policy per operation? ideally, it should be at batch level.

Reply:
In a batch, some operations can succeed while some might fail. Only failed ones need to be retried.

Reply:
That's fine, but once we see the failure, we should set the retry policy based on the type of failure. Otherwise, creating it for every operation will lead to unnecessary repetition.